File: lib/data/models/destination.dart
import 'package:tour_leader/data/models/hotel.dart';
import 'package:tour_leader/data/models/review.dart';

class Destination {
  final String id;
  final String name;
  final String description;
  final String longDescription;
  final String country;
  final String city;
  final double latitude;
  final double longitude;
  final List<String> imageUrls;
  final List<String> videoUrls;
  final double rating;
  final int reviewCount;
  final List<Review> reviews;
  final List<Hotel> hotels;
  final List<String> attractions;
  final List<String> categories;
  final double price;
  final String currency;
  final bool isPopular;
  final bool isFeatured;
  final String bestTimeToVisit;
  final List<String> activities;
  final Map<String, dynamic> additionalInfo;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Destination({
    required this.id,
    required this.name,
    required this.description,
    required this.longDescription,
    required this.country,
    required this.city,
    required this.latitude,
    required this.longitude,
    required this.imageUrls,
    this.videoUrls = const [],
    this.rating = 0.0,
    this.reviewCount = 0,
    this.reviews = const [],
    this.hotels = const [],
    this.attractions = const [],
    this.categories = const [],
    this.price = 0.0,
    this.currency = 'USD',
    this.isPopular = false,
    this.isFeatured = false,
    this.bestTimeToVisit = '',
    this.activities = const [],
    this.additionalInfo = const {},
    required this.createdAt,
    required this.updatedAt,
  });

  factory Destination.fromJson(Map<String, dynamic> json) {
    return Destination(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      longDescription: json['longDescription'] ?? '',
      country: json['country'] ?? '',
      city: json['city'] ?? '',
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      imageUrls: List<String>.from(json['imageUrls'] ?? []),
      videoUrls: List<String>.from(json['videoUrls'] ?? []),
      rating: (json['rating'] ?? 0.0).toDouble(),
      reviewCount: json['reviewCount'] ?? 0,
      reviews:
          (json['reviews'] as List?)?.map((e) => Review.fromJson(e)).toList() ??
          [],
      hotels:
          (json['hotels'] as List?)?.map((e) => Hotel.fromJson(e)).toList() ??
          [],
      attractions: List<String>.from(json['attractions'] ?? []),
      categories: List<String>.from(json['categories'] ?? []),
      price: (json['price'] ?? 0.0).toDouble(),
      currency: json['currency'] ?? 'USD',
      isPopular: json['isPopular'] ?? false,
      isFeatured: json['isFeatured'] ?? false,
      bestTimeToVisit: json['bestTimeToVisit'] ?? '',
      activities: List<String>.from(json['activities'] ?? []),
      additionalInfo: Map<String, dynamic>.from(json['additionalInfo'] ?? {}),
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String(),
      ),
      updatedAt: DateTime.parse(
        json['updatedAt'] ?? DateTime.now().toIso8601String(),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'longDescription': longDescription,
      'country': country,
      'city': city,
      'latitude': latitude,
      'longitude': longitude,
      'imageUrls': imageUrls,
      'videoUrls': videoUrls,
      'rating': rating,
      'reviewCount': reviewCount,
      'reviews': reviews.map((e) => e.toJson()).toList(),
      'hotels': hotels.map((e) => e.toJson()).toList(),
      'attractions': attractions,
      'categories': categories,
      'price': price,
      'currency': currency,
      'isPopular': isPopular,
      'isFeatured': isFeatured,
      'bestTimeToVisit': bestTimeToVisit,
      'activities': activities,
      'additionalInfo': additionalInfo,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  Destination copyWith({
    String? id,
    String? name,
    String? description,
    String? longDescription,
    String? country,
    String? city,
    double? latitude,
    double? longitude,
    List<String>? imageUrls,
    List<String>? videoUrls,
    double? rating,
    int? reviewCount,
    List<Review>? reviews,
    List<Hotel>? hotels,
    List<String>? attractions,
    List<String>? categories,
    double? price,
    String? currency,
    bool? isPopular,
    bool? isFeatured,
    String? bestTimeToVisit,
    List<String>? activities,
    Map<String, dynamic>? additionalInfo,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Destination(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      longDescription: longDescription ?? this.longDescription,
      country: country ?? this.country,
      city: city ?? this.city,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      imageUrls: imageUrls ?? this.imageUrls,
      videoUrls: videoUrls ?? this.videoUrls,
      rating: rating ?? this.rating,
      reviewCount: reviewCount ?? this.reviewCount,
      reviews: reviews ?? this.reviews,
      hotels: hotels ?? this.hotels,
      attractions: attractions ?? this.attractions,
      categories: categories ?? this.categories,
      price: price ?? this.price,
      currency: currency ?? this.currency,
      isPopular: isPopular ?? this.isPopular,
      isFeatured: isFeatured ?? this.isFeatured,
      bestTimeToVisit: bestTimeToVisit ?? this.bestTimeToVisit,
      activities: activities ?? this.activities,
      additionalInfo: additionalInfo ?? this.additionalInfo,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Destination && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'Destination(id: $id, name: $name, country: $country, city: $city)';
  }
}

File: internal/models/destination.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Destination struct {
	ID               uuid.UUID `json:"id" db:"id"`
	Name             string    `json:"name" db:"name" validate:"required,min=2,max=100"`
	Description      string    `json:"description" db:"description" validate:"required,min=10,max=500"`
	LongDescription  string    `json:"longDescription" db:"long_description" validate:"required,min=50,max=2000"`
	Country          string    `json:"country" db:"country" validate:"required,min=2,max=100"`
	City             string    `json:"city" db:"city" validate:"required,min=2,max=100"`
	Latitude         float64   `json:"latitude" db:"latitude" validate:"required,latitude"`
	Longitude        float64   `json:"longitude" db:"longitude" validate:"required,longitude"`
	ImageUrls        []string  `json:"imageUrls" db:"image_urls"`
	VideoUrls        []string  `json:"videoUrls" db:"video_urls"`
	Rating           float64   `json:"rating" db:"rating"`
	ReviewCount      int       `json:"reviewCount" db:"review_count"`
	Price            float64   `json:"price" db:"price" validate:"min=0"`
	Currency         string    `json:"currency" db:"currency" validate:"required,len=3"`
	IsPopular        bool      `json:"isPopular" db:"is_popular"`
	IsFeatured       bool      `json:"isFeatured" db:"is_featured"`
	BestTimeToVisit  string    `json:"bestTimeToVisit" db:"best_time_to_visit"`
	Activities       []string  `json:"activities" db:"activities"`
	Attractions      []string  `json:"attractions" db:"attractions"`
	Categories       []string  `json:"categories" db:"categories"`
	AdditionalInfo   string    `json:"additionalInfo" db:"additional_info"` // JSON field
	CreatedAt        time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt        time.Time `json:"updatedAt" db:"updated_at"`

	// Related data loaded separately
	Reviews []Review `json:"reviews,omitempty"`
	Hotels  []Hotel  `json:"hotels,omitempty"`
}

type DestinationRequest struct {
	Name            string   `json:"name" validate:"required,min=2,max=100"`
	Description     string   `json:"description" validate:"required,min=10,max=500"`
	LongDescription string   `json:"longDescription" validate:"required,min=50,max=2000"`
	Country         string   `json:"country" validate:"required,min=2,max=100"`
	City            string   `json:"city" validate:"required,min=2,max=100"`
	Latitude        float64  `json:"latitude" validate:"required,latitude"`
	Longitude       float64  `json:"longitude" validate:"required,longitude"`
	Price           float64  `json:"price" validate:"min=0"`
	Currency        string   `json:"currency" validate:"required,len=3"`
	BestTimeToVisit string   `json:"bestTimeToVisit"`
	Activities      []string `json:"activities"`
	Attractions     []string `json:"attractions"`
	Categories      []string `json:"categories"`
	AdditionalInfo  string   `json:"additionalInfo"`
}

type DestinationResponse struct {
	Destination
	TotalReviews int     `json:"totalReviews"`
	AvgRating    float64 `json:"avgRating"`
	HotelCount   int     `json:"hotelCount"`
	MinPrice     float64 `json:"minPrice"`
	MaxPrice     float64 `json:"maxPrice"`
}

type DestinationFilter struct {
	Country       string   `json:"country" form:"country"`
	City          string   `json:"city" form:"city"`
	Categories    []string `json:"categories" form:"categories"`
	MinPrice      float64  `json:"minPrice" form:"minPrice"`
	MaxPrice      float64  `json:"maxPrice" form:"maxPrice"`
	MinRating     float64  `json:"minRating" form:"minRating"`
	IsPopular     *bool    `json:"isPopular" form:"isPopular"`
	IsFeatured    *bool    `json:"isFeatured" form:"isFeatured"`
	Activities    []string `json:"activities" form:"activities"`
	SortBy        string   `json:"sortBy" form:"sortBy"` // price, rating, name, created_at
	SortOrder     string   `json:"sortOrder" form:"sortOrder"` // asc, desc
	Page          int      `json:"page" form:"page"`
	Limit         int      `json:"limit" form:"limit"`
	Search        string   `json:"search" form:"search"`
	NearLatitude  float64  `json:"nearLatitude" form:"nearLatitude"`
	NearLongitude float64  `json:"nearLongitude" form:"nearLongitude"`
	RadiusKm      float64  `json:"radiusKm" form:"radiusKm"`
}

func (d *Destination) BeforeCreate() {
	d.ID = uuid.New()
	d.CreatedAt = time.Now()
	d.UpdatedAt = time.Now()
}

func (d *Destination) BeforeUpdate() {
	d.UpdatedAt = time.Now()
}
